indexer/indexer_tests/osm_editor_test.cpp

```diff
@@ -0,0 +1,621 @@
+  TEST_EQUAL(created.front(), createdId.m_index, ());
+}
+
+void EditorTest::OnMapDeregisteredTest()
+{
+  auto & editor = osm::Editor::Instance();
+
+  TestCity london(m2::PointD(1.0, 1.0), "London", "en", 100 /* rank */);
+  TestCity moscow(m2::PointD(2.0, 2.0), "Moscow", "ru", 100 /* rank */);
+  BuildMwm("TestWorld", feature::DataHeader::world,[&london, &moscow](TestMwmBuilder & builder)
+  {
+    builder.Add(london);
+    builder.Add(moscow);
+  });
+
+  TestCafe cafeLondon(m2::PointD(1.0, 1.0), "London Cafe", "en");
+  auto const gbMwmId = BuildMwm("GB", feature::DataHeader::country, [&cafeLondon](TestMwmBuilder & builder)
+  {
+    builder.Add(cafeLondon);
+  });
+  ASSERT(gbMwmId.GetInfo(), ());
+
+  TestCafe cafeMoscow(m2::PointD(2.0, 2.0), "Moscow Cafe", "ru");
+  auto const rfMwmId = BuildMwm("RF", feature::DataHeader::country, [&cafeMoscow](TestMwmBuilder & builder)
+  {
+    builder.Add(cafeMoscow);
+  });
+  ASSERT(rfMwmId.GetInfo(), ());
+
+  ForEachCafeAtPoint(m_model, m2::PointD(1.0, 1.0), [&editor](FeatureType & ft)
+  {
+    osm::EditableMapObject emo;
+    FillEditableMapObject(editor, ft, emo);
+    emo.SetBuildingLevels("1");
+    TEST_EQUAL(editor.SaveEditedFeature(emo), osm::Editor::SaveResult::SavedSuccessfully, ());
+  });
+
+  ForEachCafeAtPoint(m_model, m2::PointD(2.0, 2.0), [&editor](FeatureType & ft)
+  {
+    osm::EditableMapObject emo;
+    FillEditableMapObject(editor, ft, emo);
+    emo.SetBuildingLevels("1");
+    TEST_EQUAL(editor.SaveEditedFeature(emo), osm::Editor::SaveResult::SavedSuccessfully, ());
+  });
+
+  TEST(!editor.m_features.empty(), ());
+  TEST_EQUAL(editor.m_features.size(), 2, ());
+
+  editor.OnMapDeregistered(gbMwmId.GetInfo()->GetLocalFile());
+
+  TEST_EQUAL(editor.m_features.size(), 1, ());
+  auto const editedMwmId = editor.m_features.find(rfMwmId);
+  bool result = (editedMwmId != editor.m_features.end());
+  TEST(result, ());
+}
+
+void EditorTest::Cleanup(platform::LocalCountryFile const & map)
+{
+  platform::CountryIndexes::DeleteFromDisk(map);
+  map.DeleteFromDisk(MapOptions::Map);
+}
+
+void EditorTest::InitEditorForTest()
+{
+  auto & editor = osm::Editor::Instance();
+
+  editor.m_storage = make_unique<editor::StorageMemory>();
+  editor.ClearAllLocalEdits();
+
+  editor.SetMwmIdByNameAndVersionFn([this](string const & name) -> MwmSet::MwmId
+  {
+    return m_model.GetIndex().GetMwmIdByCountryFile(platform::CountryFile(name));
+  });
+
+  editor.SetFeatureLoaderFn([this](FeatureID const & fid) -> unique_ptr<FeatureType>
+  {
+    unique_ptr<FeatureType> feature(new FeatureType());
+    Index::FeaturesLoaderGuard const guard(m_model.GetIndex(), fid.m_mwmId);
+    if (!guard.GetOriginalFeatureByIndex(fid.m_index, *feature))
+      return nullptr;
+    feature->ParseEverything();
+    return feature;
+  });
+
+  editor.SetFeatureOriginalStreetFn([this](FeatureType & ft) -> string
+  {
+    search::ReverseGeocoder const coder(m_model.GetIndex());
```

Comment:
Arsentiy Milchakov <notifications@github.com> writes:

> In indexer/indexer_tests/osm_editor_test.cpp:
>
>> +    return m_model.GetIndex().GetMwmIdByCountryFile(platform::CountryFile(name));
>> +  });
>> +
>> +  editor.SetFeatureLoaderFn([this](FeatureID const & fid) -> unique_ptr<FeatureType>
>> +  {
>> +    unique_ptr<FeatureType> feature(new FeatureType());
>> +    Index::FeaturesLoaderGuard const guard(m_model.GetIndex(), fid.m_mwmId);
>> +    if (!guard.GetOriginalFeatureByIndex(fid.m_index, *feature))
>> +      return nullptr;
>> +    feature->ParseEverything();
>> +    return feature;
>> +  });
>> +
>> +  editor.SetFeatureOriginalStreetFn([this](FeatureType & ft) -> string
>> +  {
>> +    search::ReverseGeocoder const coder(m_model.GetIndex());
>
> это легкий класс, содержит только ссылку на индекс, в конструкторе
> только она и инициализируется

OK.

>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.*
>

-- 
Best regards, Yuri